### PR TITLE
ENH Make year and quarter dash-separatable

### LINKED ISSUE #9688: ENH: make period_range parse strings like '2012-Q1', not just '2012Q1'
This format is used, e.g., in datasets from the ECB. 

Code example:

In [25]: PD.period_range(start = '2012Q1', periods = 4, freq = 'Q')
Out[25]:
<class 'pandas.tseries.period.PeriodIndex'>
[2012Q1, ..., 2012Q4]
Length: 4, Freq: Q-DEC

In [26]: PD.period_range(start = '2012-Q1', periods = 4, freq = 'Q')
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
C:\Users\stefan\Anaconda\envs\py343\lib\site-packages\pandas\tseries\tools.py in
 parse_time_string(arg, freq, dayfirst, yearfirst)
    463         parsed, reso = dateutil_parse(arg, default, dayfirst=dayfirst,
--> 464                                       yearfirst=yearfirst)
    465     except Exception as e:

C:\Users\stefan\Anaconda\envs\py343\lib\site-packages\pandas\tseries\tools.py in
 dateutil_parse(timestr, default, ignoretz, tzinfos, **kwargs)
    489     if res is None:
--> 490         raise ValueError("unknown string format")
    491

ValueError: unknown string format

During handling of the above exception, another exception occurred:

DateParseError                            Traceback (most recent call last)
<ipython-input-26-a960a544a087> in <module>()
----> 1 PD.period_range(start = '2012-Q1', periods = 4, freq = 'Q')

C:\Users\stefan\Anaconda\envs\py343\lib\site-packages\pandas\tseries\period.py i
n period_range(start, end, periods, freq, name)
   1433     """
   1434     return PeriodIndex(start=start, end=end, periods=periods,
-> 1435                        freq=freq, name=name)

C:\Users\stefan\Anaconda\envs\py343\lib\site-packages\pandas\tseries\period.py i
n __new__(cls, data, ordinal, freq, start, end, periods, copy, name, tz, **kwarg
s)
    637             else:
    638                 data, freq = cls._generate_range(start, end, periods,
--> 639                                                  freq, kwargs)
    640         else:
    641             ordinal, freq = cls._from_arraylike(data, freq, tz)

C:\Users\stefan\Anaconda\envs\py343\lib\site-packages\pandas\tseries\period.py i
n _generate_range(cls, start, end, periods, freq, fields)
    651                 raise ValueError('Can either instantiate from fields '
    652                                  'or endpoints, but not both')
--> 653             subarr, freq = _get_ordinal_range(start, end, periods, freq)

    654         elif field_count > 0:
    655             subarr, freq = _range_from_fields(freq=freq, **fields)

C:\Users\stefan\Anaconda\envs\py343\lib\site-packages\pandas\tseries\period.py i
n _get_ordinal_range(start, end, periods, freq)
   1281
   1282     if start is not None:
-> 1283         start = Period(start, freq)
   1284     if end is not None:
   1285         end = Period(end, freq)

C:\Users\stefan\Anaconda\envs\py343\lib\site-packages\pandas\tseries\period.py i
n __init__(self, value, freq, ordinal, year, month, quarter, day, hour, minute,
second)
    124             value = value.upper()
    125
--> 126             dt, _, reso = parse_time_string(value, freq)
    127             if freq is None:
    128                 try:

C:\Users\stefan\Anaconda\envs\py343\lib\site-packages\pandas\tseries\tools.py in
 parse_time_string(arg, freq, dayfirst, yearfirst)
    465     except Exception as e:
    466         # TODO: allow raise of errors within instead
--> 467         raise DateParseError(e)
    468
    469     if parsed is None:

DateParseError: unknown string format

### PR DESCRIPTION
**Pull request by williamhogman**
This closes #9688 by making years and quarters separable by dashes.

### CODE REVIEW

File: pandas/tseries/tests/test_offsets.py [hunk 1]
```diff
@@ -3018,6 +3018,21 @@ def test_parse_time_string():
     assert reso == reso_lower
 
 
+def test_parse_time_quarter_w_dash():
+    pairs = [
```
Comment (jreback) [inline pandas/tseries/tests/test_offsets.py]: add the issue in a comment here

File: pandas/tseries/tests/test_offsets.py [hunk 2]
```diff
@@ -3018,6 +3018,21 @@ def test_parse_time_string():
...
+        ('1988-Q2', '1988Q2'),
+        ('2Q-1988', '2Q1988'),
+    ]
+
+    for dashed, normal in pairs:
+        (date_dash, parsed_dash, reso_dash) = parse_time_string(dashed)
+        (date, parsed, reso) = parse_time_string(normal)
+
+        assert date_dash == date
+        assert parsed_dash == parsed
```
Comment (jreback) [inline pandas/tseries/tests/test_offsets.py]: use ``self.assertEqual(...)`` or ``self.assertTrue(...)``

File: pandas/tseries/tests/test_offsets.py (hunk 1 above)
Comment (jreback) [inline pandas/tseries/tests/test_offsets.py]: are there possibilities that it *shouldn't* match?

**Comment by williamhogman** [comment]
@jreback I think I've adjusted them accordingly, although I'm not quite sure about the how you want the test-case class. Anyhow, tell me if it looks OK so I can rebase it :)

File: pandas/tseries/tests/test_offsets.py [hunk 3]
```diff
@@ -3010,12 +3010,33 @@ def test_get_offset():
...
 
 
-def test_parse_time_string():
-    (date, parsed, reso) = parse_time_string('4Q1984')
-    (date_lower, parsed_lower, reso_lower) = parse_time_string('4q1984')
-    assert date == date_lower
-    assert parsed == parsed_lower
-    assert reso == reso_lower
+class TestParseTimeString(tm.TestCase):
+
```
Comment (jreback) [inline pandas/tseries/tests/test_offsets.py]: add the issue number as a comment here

File: doc/source/whatsnew/v0.16.1.txt [hunk 4]
```diff
@@ -22,6 +22,8 @@ API changes
 Enhancements
 ~~~~~~~~~~~~
 
+- Add support for separating years and quarters using dashes  (:issue:`9688`)
+
```
Comment (jreback) [inline doc/source/whatsnew/v0.16.1.txt]: maybe say parsing here to Periods's (and give an example, e.g '2012-Q1')

File: pandas/tseries/tests/test_offsets.py [hunk 5]
```diff
@@ -3010,12 +3010,33 @@ def test_get_offset():
...
+        self.assertEqual(parsed, parsed_lower)
+        self.assertEqual(reso, reso_lower)
+
+    def test_parse_time_quarter_w_dash(self):
+        pairs = [
+            ('1988-Q2', '1988Q2'),
+            ('2Q-1988', '2Q1988'),
+        ]
+
+        # https://github.com/pydata/pandas/issues/9688
```
Comment (jreback) [inline pandas/tseries/tests/test_offsets.py]: ignore my previous, just move this to the top of the function

**Comment by jreback** [comment]
minor comments, pls squash as well. thanks. ping when green.

**Comment by williamhogman** [comment]
@jreback That should do it, commits are now squashed.

**Comment by jreback** [comment]
thank you!

**Comment by williamhogman** [comment]
Happy to help, pandas has saved my ass too many times to count.

William Rudenmalm
tel. +46 70 263 86 33
www.whn.se

**Pull request merged**